bluesky/plan_stubs.py

```diff
@@ -295,6 +302,91 @@ def inner_mvr():
 movr = mvr  # synonym
 
 
+def rd(obj, *, default_value=0):
+    """Reads a single-value non-triggered object
+
+    This is a helper plan to get the scalar value out of a Device (such as an EpicsMotor or a
+    single EpicsSignal).
+
+    For devices that have more than one read key the following rules are used:
+
+    - if exactly 1 field is hinted that value is used
+    - if no fields are hinted and there is exactly 1 value in the
+      reading that value is used
+    - if more than one field is hinted an Exception is raised
+    - if no fields are hinted and there is more than one key in the reading an
+      Exception is raised
+
+    The devices is not triggered and this plan does not create any Events
+
+    Parameters
+    ----------
+    obj : Device
+        The device to be read
+
+    default_value : Any
+        The value to return when not running in a "live" RunEngine (when
+        ret = yield from obj.read() returns None)
```

Comment:
I'm not sure I understand what `default_value` does here. I wasn't aware that `read` behaves differently outside of a run engine. I haven't been following `bluesky` very closely so I suspect I've missed something.

Reply:
Oh, don't worry, this isn't new at all. It's just that when you do an adaptive pattern like this:

```py
def plan(...):
    ...
    reading = yield from bluesky.plan_stubs.read(...)
    ...
```

and then iterative over it like `list(plan)` then `reading` is `None`. Executing it with the `RunEngine` (or hypothetically anything that calls `send` to send a value back) gives you an actual reading.

Reply:
I see. I don't quite understand why `reading` becomes `None` in this case instead of a `Msg` object but I do understand why a value is needed.

Reply:
Either way, the docstring should have the example from your comment instead of the line it currently has, since `yield from obj.read()` is probably some sort of error in all cases (yielding from a dictionary)

Reply:
Good catch on the docstring.

The `reading` would never be a `Msg` object. Going step by step: the `Msg('read', ...)` yielded by `bluesky.plan_stubs.read(...)` on the right side of `=` goes to the RE, is dispatched by the command lookup table to `RunEngine._read(msg)`, which calls `reading = obj.read()`, and then calls `plan.send(reading)` to send that `reading` back to the plan, where it lands on the left side of the `=`.

If the plan is consumed by something that does not `.send`, then the left side of the `=` just gets `None`.

This only comes up in the context of adaptive plans. (Many plans do not capture the result of `yield from`.) The `None` can be interpreted as, “Whatever is consuming this plan apparently isn’t actually getting readings from hardware. It might be a simulator or summarizer. We can’t actually do adaptive logic without readings, so here is `None` as a placeholder.” This behavior has been in place since early days.

Am I addressing the point of confusion? Apologies if I’m talking past you.

Reply:
oops, that is definitely a typo in the docs, sorry.

Reply:
To illustrate my confusion, here is an IPython session I just ran using your example:
```
In [1]: import bluesky.plan_stubs                                               

In [2]: def plan(): 
   ...:     reading = yield from bluesky.plan_stubs.read(object) 
   ...:     return reading 
   ...:                                                                         

In [3]: list(plan())                                                            
Out[3]: [read: (<class 'object'>), (), {}]
```
What am I missing?

Reply:
I had a misunderstanding about how `yield from` and generators worked. See the source of confusion:
```
In [1]: import bluesky.plan_stubs                                               

In [2]: def plan(): 
   ...:     reading = yield from bluesky.plan_stubs.read(object) 
   ...:     print(reading) 
   ...:     yield reading 
   ...:                                                                         

In [3]: list(plan())                                                            
None
Out[3]: [read: (<class 'object'>), (), {}, None]
```

Reply:
Oh, that is confusing. One thing this tells as, as an aside, in that our less "special" `Msg.__repr__` is an improvement in situations like this. On bluesky `master` this looks like:

```
In [5]: def plan(): 
   ...:     reading = yield from bluesky.plan_stubs.read(object) 
   ...:     print('reading', reading) 
   ...:     yield reading 
   ...:                                                                                                                                                                                       

In [6]: list(plan())                                                                                                                                                                          
reading None
Out[6]: [Msg('read', obj=<class 'object'>, args=(), kwargs={}, run=None), None]
```